python/pyspark/sql/worker/plan_data_source_read.py

```diff
@@ -54,17 +54,31 @@
 
 
 def records_to_arrow_batches(
-    output_iter: Iterator[Tuple],
+    output_iter: Union[Iterator[Tuple], Iterator[pa.RecordBatch]],
     max_arrow_batch_size: int,
     return_type: StructType,
     data_source: DataSource,
 ) -> Iterable[pa.RecordBatch]:
     """
-    Convert an iterator of Python tuples to an iterator of pyarrow record batches.
+    First check if the iterator yields pyarrow record batched, if so, yield them directly.
+    Otherwise, convert an iterator of Python tuples to an iterator of pyarrow record batches.
+    For each Python tuple, check the types of each field and append it to the records batch.
+    """
 
-    For each python tuple, check the types of each field and append it to the records batch.
+    # Create a copy of the iterator using itertools.tee
+    output_iter, output_iter_copy = tee(output_iter)
 
-    """
+    try:
+        # Check if the first element is of type pa.RecordBatch
+        # In that case, yield all elements and return
+        first_element = next(output_iter_copy)
+        if isinstance(first_element, pa.RecordBatch):
+            yield first_element
+            for element in output_iter_copy:
+                yield element
```

Comment:
I think we need to check if the schema of the arrow batch  (# of output columns, data types, etc)  matches the expected schema of the data source (similar to the checks below for iterator[row])

Reply:
Indeed. I have added a check on # of output columns and column names for the moment. Probably data types should be done too.